### DISCUSSION TITLE
Adding own "dictionary" or phrase pairs

### DISCUSSION THREAD

**Discussion by mirozahorak** [ggml-org/whisper.cpp#602] [Ideas]
First of all, amazing work @ggerganov & the rest!
I wonder if the models include translating from  AI vocabulary :)  I am programmer myself,  but when i read your discussion it is like a foreign language to me. Just kidding, no need to implement :)

To my point: Is there a chance of adding my own phrase pairs or word pairs so that transcript precision for certain specific topics can be enhanced? Like names or phrases from specialised areas. From my tests I have noticed that this is area of most inaccurate transcripts.

The same applies to translating, but that is more complex topic.

**Comment by RndyP** [comment]
gerganov has the command example that does this. I'm in the same boat as you. You can create a custom dictionary. I had it up and running. The problem was false positives. When selecting from a dictionary it will choose the highest probability phrase. The problem is that the probability value is all over the place sometimes, and I couldn't depend on it. gerganov has this problem on his radar, but is swamped with other things. I can't find the issue # right now.

Anyway, what I did to get my preferred words to work is create an error dictionary. I "train" it by saying the word over and over and adding all the incorrect words to the dictionary.

**Reply by mirozahorak** [reply]
Thanks for the insight. Error dictionary solution is something I have not heard about, neither I knew that i can actually train the model myself. 
I need to search more in the repo for this example. 

To avoid false positives – maybe there is a way how to increase the probability by percentage, relatively?

**Reply by RndyP** [reply]
I think what you are looking for is in between regular inference, and a special "command" dictionary. You want to favor certain words. I'm thinking to do that, you might want a custom ggml*.bin file. Whisper is a weakly trained, large dataset, that has every word under the sun in it. Would be nice if there could be filtering, like, no names adjectives, scrabble words, etc.....

**Comment by edshamis** [comment]
Any plans for dictionary?

**Comment by RndyP** [comment]
Here's a small section of my error dictionary. I have about 120 words when I'm in command mode. Sometimes I go into what I call "raw" mode, where I don't correct.

vector<pair<CString,vector<CString>>>	m_vDictionary;


CDictionary::CDictionary()
{
	m_vDictionary = {
		
		{{"about"},
		{"aboat","abolt","theboat","avault","abulk","thebolt"}},

		{{"add"},
		{"air","aired","ed","aad","head","ithad"}},

		{{"acquisition"},
		{"acquisitions"}},

		{{"alarm"},
		{"alarms","arm","alert","thealarm","warm","worm","alarmed","aworm","alarmani","learned","tolearn"
		,"iam","larm","blurm"}},

		{{"analysis"},
		{}},

		{{"anchor"},
		{"acre","baker","panker","anger","ager","sanker","hager","egger"}},

		{{"artificial"},
		{"articical","hurtofficial"}},

		{{"backup"},
		{"back"}},


/***************************************************************************
correct a word via sub class dictionary
ErrorCorrected is not touched if not corrected or failed
***************************************************************************/
bool CDictionary::Correct(char* word,CORRECTION_STATUS& ErrorCorrected,CString& str)
{
	if ( *word==0 )
		return false;
	if ( IsNumeric(word) )
		return true;
	// is there an exact match?
	for (const auto& entry : m_vDictionary)
	{
		if (entry.first==word)
			return true;
	}
	// no exact match - look for same words in error
	for (const auto& entry : m_vDictionary)
	{
		for (const auto& error : entry.second)
		{
			if (error==word)
			{
				str=word;
				ErrorCorrected=CORRECTED;
				strcpy_s(word,MaxStringSize,entry.first.GetString());
				return true;
			}
		}
	}
	str=word;
	ErrorCorrected=FAILED;
	return false;
}



This has got command mode working pretty well. There are certain words that give Whisper a lot of trouble, and I have to avoid them. Most have a few errors words, and some are reliable and don't need any correction (like "analysis"), If I get more than about 20 error words, I give up on that word and think of something similar.
